File: JavaScript/src/NyzoStringPrefilledData.js
// https://github.com/n-y-z-o/nyzoVerifier/blob/master/src/main/java/co/nyzo/verifier/nyzoString/NyzoStringPrefilledData.java

const { NyzoString } = require('./NyzoString')
// JS does not support native int64
const Int64 = require('node-int64')


class NyzoStringPrefilledData extends NyzoString {

  constructor(receiverIdentifier, senderData, amount) {
    // Amount is expected as Int64
    senderData = senderData.slice(0, 32)
    let bufferLen = 32 + 1 + senderData.length
    if (amount == null) {
        amount = new Int64(0)
    }
    if (amount > 0) {
        bufferLen += 8
    }
    let bytes = Buffer.alloc(bufferLen)
    Buffer.from(receiverIdentifier).copy(bytes, 0)
    bytes[32] = senderData.length
    Buffer.from(senderData).copy(bytes, 33)
    if (amount > 0) {
        bytes[32] |= 0b10000000
        amount.copy(bytes, 33 + senderData.length)
    }
    super('pre_', bytes)
    this.receiverIdentifier = receiverIdentifier
    this.senderData = senderData
    this.amount = amount
  }

  getReceiverIdentifier() {
        return this.receiverIdentifier
  }

  getSenderData() {
        return this.senderData
  }

  getAmount() {
        // This returns an Int64
        return this.amount
  }

  static fromByteBuffer(buffer) {
    const receiverBuffer = buffer.slice(0, 32)
    const senderDataLength = Math.min(buffer[32] & 0b00111111, 32)
    const dataBuffer = buffer.slice(33, 33 + senderDataLength)
    let amount = new Int64(0)
    if ((buffer[32] & 0b10000000) == 0b10000000) {
        // Amount was encoded
        amount = new Int64(buffer.slice(33 + senderDataLength, 33 + senderDataLength + 8))
    }
    return new NyzoStringPrefilledData(receiverBuffer, dataBuffer, amount)
  }

  static fromHex(receiverHexString, dataHexString, amount) {
    // Amount is expected as Int64
    const filteredString = receiverHexString.split('-').join('').slice(0, 64)
    const receiverBuffer = Buffer.from(filteredString, 'hex')
    const dataBuffer = Buffer.from(dataHexString, 'hex')
    return new NyzoStringPrefilledData(receiverBuffer, dataBuffer, amount)
  }

}


module.exports = {
    version: "0.0.2",
    NyzoStringPrefilledData
}

File: Python/nyzostrings/nyzostrings/nyzostringprefilleddata.py
"""
Nyzo String for Prefilled Data
"""

from nyzostrings.nyzostring import NyzoString


__version__ = "0.0.2"


class NyzoStringPrefilledData(NyzoString):
    def __init__(
        self,
        receiver_identifier: bytes,
        sender_data: bytes,
        amount: int = 0  # Unit is integer, micronyzo
    ) -> None:
        sender_data = sender_data[:32]
        sender_data_len = len(sender_data)
        buf_len = 32 + 1 + sender_data_len
        if amount > 0:
            # Add amount
            buf_len += 8
        bytes_content = bytearray(buf_len)
        bytes_content[0:32] = receiver_identifier
        bytes_content[32] = sender_data_len
        bytes_content[33: 33 + sender_data_len] = sender_data
        if amount > 0:
            bytes_content[32] |= 0b10000000
            bytes_content[33 + sender_data_len: 33 + sender_data_len + 8] = amount.to_bytes(8, byteorder="big")
        super().__init__("pre_", bytes_content)
        self.receiver_identifier = receiver_identifier
        self.sender_data = sender_data
        self.amount = amount

    @classmethod
    def from_bytes(cls, byte_buffer: bytes) -> "NyzoStringPrefilledData":
        buffer = memoryview(byte_buffer)
        receiver_buffer = buffer[:32]
        amount = 0
        sender_data_length = min(buffer[32] & 0b00111111, 32)
        data_buffer = buffer[33: 33 + sender_data_length]
        if buffer[32] & 0b10000000 == 0b10000000:
            # Amount was encoded
            amount = int.from_bytes(
                buffer[33 + sender_data_length: 33 + sender_data_length + 8], byteorder="big", signed=False
            )
        return NyzoStringPrefilledData(
            receiver_buffer,
            data_buffer,
            amount
        )

    @classmethod
    def from_hex(
        cls,
        receiver_identifier_hex: str,
        sender_data_hex: str,
        amount: int=0
    ) -> "NyzoStringPrefilledData":
        filtered_string = receiver_identifier_hex.replace("-", "")[:64]
        receiver = bytes.fromhex(filtered_string)
        filtered_string = sender_data_hex.replace("-", "")[:64]
        data = bytes.fromhex(filtered_string)
        return NyzoStringPrefilledData(receiver, data, amount)
